_posts/2019-04-24-istio-service-mesh-s2s.md

```diff
@@ -0,0 +1,431 @@
+
+<img class="image fit" style="margin:0px auto; max-width: 446px;" src="{{ '/img/2019-04-14-istio-service-mesh-s2s/istio-basic-setup.png' | prepend: site.baseurl }}" alt="Basic setup in Istio" />
+
+### Enabling mutual TLS (mTLS)
+
+Currently the service can connect to the backend just fine. 
+TLS is currently not used to communicate between the service. 
+
+The following manifest defines a policy which changes this. 
+It is a namespace scoped policy telling Istio that all services in the `with-istio` namespace should **ONLY** accept mTLS connections.
+This configuration will be picked up by Pilot and distributed to all Envoy proxies in the `with-istio` namespace.
+When this policy is applied, Envoy will drop any requests it gets that don't use mTLS.
+
+```yaml
+apiVersion: "authentication.istio.io/v1alpha1"
+kind: "Policy"
+metadata:
+  name: "default"
+  namespace: with-istio
+spec:
+  peers:
+  - mtls: 
+      mode: STRICT
+```
+
+Note, this policy only affects the incoming connections on the Envoy proxy.
+When a request would be sent to the test-app service now, it would be rejected with an HTTP 503 error code.
+This is shown in the following drawing.
+
+<img class="image fit" style="margin:0px auto; max-width: 408px;" src="{{ '/img/2019-04-14-istio-service-mesh-s2s/istio-mtls-broken.png' | prepend: site.baseurl }}" alt="Broken mTLS drawing" />
+
+Next, the outgoing (client) connections needs to be configured to use mTLS.
+This can be done by specifying a destination rule for the services.
+A destination rule defines a set of rules that are evaluated for every outgoing request from a proxy.
+This rules defines that every proxy in `with-istio` namespace needs to use mutual TLS for every service that ends with `.local`.
+By applying this rule, the requests will succeed again on the test-app service.
+
+```yaml
+apiVersion: "networking.istio.io/v1alpha3"
+kind: "DestinationRule"
+metadata:
+  name: "default"
+  namespace: with-istio
+spec:
+  host: "*.local"
+  trafficPolicy:
+    tls:
+      mode: ISTIO_MUTUAL
+```
+
+### Enabling Role Based Access Control (RBAC) on the services
+
+Services can now communicate securely over mTLS.
+To increase the security even further, RBAC can be added to the services.
+RBAC allows for roles to be defined that specify access to specific services in the cluster. 
+By attaching these roles to service accounts (which are connected to services) services can be premitted to access specific other services. 
+This limits the reach a single service has in the cluster and therefor adheres to the least privileges principle.
+
+The following manifest defines an cluster RBAC configuration.
+Only one such a configuration can exist in the entire service mesh and it needs to have the name `default`.
+The mode `ON_WITH_INCLUSION` specifies that all subjects that are listed in the inclusion section need to have RBAC enabled. 
+These subject can be namespace and/or specific services.
```

Comment:
These *subjects* can be *namespaces* and/or specific services.

Reply:
Updated